plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,515 @@
+#define FU_PXI_DEVICE_OTA_BUF_SZ		32	/* bytes */
+
+struct _FuPxiDevice {
+	FuUdevDevice	 parent_instance;
+	guint8		 status;
+	guint8		 new_flow;
+	guint16		 offset;
+	guint16		 checksum;
+	guint32		 max_object_size;
+	guint16		 mtu_size;
+	guint16		 prn_threshold;
+	guint8		 spec_check_result;
+};
+
+G_DEFINE_TYPE (FuPxiDevice, fu_pxi_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_pxi_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuPxiDevice *self = FU_PXI_DEVICE (device);
+	fu_common_string_append_kx (str, idt, "Status", self->status);
+	fu_common_string_append_kx (str, idt, "NewFlow", self->new_flow);
+	fu_common_string_append_kx (str, idt, "CurrentObjectOffset", self->offset);
+	fu_common_string_append_kx (str, idt, "CurrentChecksum", self->checksum);
+	fu_common_string_append_kx (str, idt, "MaxObjectSize", self->max_object_size);
+	fu_common_string_append_kx (str, idt, "MtuSize", self->mtu_size);
+	fu_common_string_append_kx (str, idt, "PacketReceiptNotificationThreshold", self->prn_threshold);
+	fu_common_string_append_kx (str, idt, "SpecCheckResult", self->spec_check_result);
+}
+
+static FuFirmware *
+fu_pxi_device_prepare_firmware (FuDevice *device,
+				GBytes *fw,
+				FwupdInstallFlags flags,
+				GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_pxi_rf_firmware_new ();
+	if (!fu_firmware_parse (firmware, fw, flags, error))
+		return NULL;
+	return g_steal_pointer (&firmware);
+}
+
+#if 0
+static gboolean
+fu_pxi_device_set_feature (FuPxiDevice *self, const guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "SetFeature", data, datasz);
+	return fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				     HIDIOCSFEATURE(datasz), (guint8 *) data,
+				     NULL, error);
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+#endif
+
+static gboolean
+fu_pxi_device_get_feature (FuPxiDevice *self, guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(datasz), data,
+				   NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "GetFeature", data, datasz);
+	return TRUE;
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static guint16
+fu_pxi_device_calculate_checksum (const guint8 *data, gsize sz)
+{
+	guint16 checksum = 0;
+	for (gsize idx = 0; idx < sz; idx++)
+		checksum += (guint16) data[idx];
+	return checksum;
+}
+
+/* FIXME: is @port required? */
+static gboolean
+fu_pxi_device_wait_notify (FuPxiDevice *self,
+			   goffset port,
+			   guint8 *status,
+			   guint16 *checksum,
+			   GError **error)
```

Comment:
@jyu99 can you answer the FIXME please

Reply:
Ok, I will check again base this commitm thank you

Reply:
I *think* the port can be zero in all cases, but didn't want to assume.

Reply:
I want reserve it because current device is BLE, but we will support other USB HID.

Reply:
No problem; we can just remove the FIXME then.